File: Java/src/NapakalakiGame/NumericBadConsequence.java
/*
 * Autores:
 * Guillermo Montes Martos
 * Manuel Ruiz López
 *
 */
package NapakalakiGame;

import java.util.ArrayList;

public class NumericBadConsequence extends BadConsequence {
    private int nVisibleTreasures = 0;
    private int nHiddenTreasures = 0;
    
    public NumericBadConsequence (String t, int l, int nVisible, int nHidden) {
            super(t,l);
            nVisibleTreasures = nVisible;
            nHiddenTreasures = nHidden;
    }
    
    public int getNVisibleTreasures(){
        return nVisibleTreasures;
    }
    
    public int getNHiddenTreasures(){
        return nHiddenTreasures;
    }
    
    @Override
    public boolean isEmpty() {
        return nVisibleTreasures==0 && nHiddenTreasures==0;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t) {
        if(nVisibleTreasures>0)
            nVisibleTreasures--;
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        if(nHiddenTreasures>0)
            nHiddenTreasures--;
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h) {
        BadConsequence nuevo;
        
        int nv = (nVisibleTreasures<v.size())? nVisibleTreasures : v.size();
        int nh = (nHiddenTreasures<h.size())? nHiddenTreasures : h.size();
        nuevo = new NumericBadConsequence(this.getText(), this.getLevels(), nv, nh);
                
        return nuevo;
    }

}

File: Ruby/lib/NumericBadConsequence.rb
#encoding: utf-8

# Autores:
# 
#     Guillermo Montes Martos
#     Manuel Ruiz López
#

require_relative 'BadConsequence.rb'

module NapakalakiGame
  class NumericBadConsequence < BadConsequence
    
    attr_reader :nVisibleTreasures
    attr_reader :nHiddenTreasures
    
    public_class_method:new
    
    def initialize(t, l, nVisible, nHidden)
      super(t, l)
      @nVisibleTreasures = nVisible
      @nHiddenTreasures = nHidden
    end

    
    def isEmpty
      @nVisibleTreasures == 0 && @nHiddenTreasures == 0
    end
    
    def substractVisibleTreasure(t)
      if @nVisibleTreasures > 0
        @nVisibleTreasures -= 1
      end
    end
    
    def substractHiddenTreasure(t)
      if @nHiddenTreasures > 0
        @nHiddenTreasures -= 1
      end
    end
    
    def adjustToFitTreasureLists(v, h)
      bd = nil
      minV = v.size > @nVisibleTreasures? @nVisibleTreasures : v.size
      minH = h.size > @nHiddenTreasures? @nHiddenTreasures : h.size
      bd = NumericBadConsequence.new(@text, @levels, minV, minH)
    end
  
  end
end